sdk/python/kfp/components/_components.py

```diff
@@ -180,18 +180,27 @@ def _react_to_incompatible_reference_type(
     raise TypeError(message)
 
 
-def _create_task_spec_from_component_and_arguments(
-    component_spec: ComponentSpec,
+def _create_task_spec_from_task_spec_and_arguments(
+    task_spec: TaskSpec,
     arguments: Mapping[str, Any],
-    component_ref: ComponentReference = None,
 ) -> TaskSpec:
-    """Constructs a TaskSpec object from component reference and arguments.
-    The function also checks the arguments types and serializes them."""
-    if component_ref is None:
-        component_ref = ComponentReference(spec=component_spec)
-    else:
-        component_ref = copy.copy(component_ref)
-        component_ref.spec = component_spec
+    """Constructs a task object from an argument-less TaskSpec and arguments.
+    
+    This function provides a reference implementation for the _default_container_task_constructor that returns TaskSpec objects.
+    The only reference-type arguments that TaskSpec can hold are TaskOutputArgument and GraphInputArgument.
+    When bridging with other frameworks, an alternative implementation should be provided that can process reference-type arguments that are native to that framework.
```

Comment:
Do you mind elaborating on what other frameworks might be?

Reply:
By other framework I mean any orchestration framework whose DSL is not build on top of `TaskSpec` and `TaskOutputReference`.
The main examples are:
* KFP: Uses `ContainerOp` and `PipelineParam`.  Implementation of _container_task_constructor for KFP:  [link](https://github.com/kubeflow/pipelines/blob/b22a0b2b35a547423ea7e57562abc42fab367de5/sdk/python/kfp/dsl/_component_bridge.py#L23)
* TFX: Uses `BaseComponent` and `Channel`. Implementation of _container_task_constructor for TFX:  [link](https://github.com/tensorflow/tfx/blob/84afa71c40418e2e9be971de452a7921d98479c0/tfx/dsl/components/bridge.py#L316)
* Airflow: Uses `Operator`. No implementations.

The reference implementation is used for framework-agnostic testing and also for creation of graph components.

Reply:
Thanks.

Just a side note. IIUC the proposed TFX component authoring user experience is: python function -> ComponentSpec + ExecutorContainerSpec, then ExecutorContainerSpec will be translated to pod manifest directly in K8Slauncher/DockerComponentLauncher, right? Are we going to change this path? I would like to see how the task spec is involved there.

Reply:
> Just a side note. IIUC the proposed TFX component authoring user experience is: python function -> ComponentSpec + ExecutorContainerSpec, then ExecutorContainerSpec will be translated to pod manifest directly in K8Slauncher/DockerComponentLauncher, right? Are we going to change this path? I would like to see how the task spec is involved there.

This was the initial idea. However I later understood that there are some task configuration options that are not part of the component, but need to be delivered to the task-like object and made available to the launcher. For example, we want the Kubernetes options and caching/retry options to reach ContainerOp in KFP and we want `KubernetesComponentConfig` to reach the `KubernetesLauncher`/`DockerComponentLauncher` in TFX.
Example KFP issue:  https://github.com/kubeflow/pipelines/issues/2942

So the we change the flow from:
python func --> ComponentSpec --> BaseComponent class with ExecutorContainerSpec -- (pass arguments)-->  BaseComponent object with ExecutorContainerSpec
to
python func --> ComponentSpec --> factory function --(pass arguments)--> TaskSpec (includes ComponentSpec) + arguments --> BaseComponent object with ExecutorContainerSpec and KubernetesComponentConfig

In a sense, we're switching from
ComponentSpec --> task-like object
to
TaskSpec --> task-like object
to preserve and pass all options.

>I would like to see how the task spec is involved there.

Let's go backwards from the end goal:
1. We want the launcher to receive KubernetesComponentConfig or caching_strategy or retry_strategy.
2. The launcher launches a task-like object (ContainerOp object or BaseComponent object).
3. Whatever function creates the task-like objects (`_default_container_task_constructor`) needs to have access to the task configuration (kubernetes, retry, caching, etc).
4. A class that stores those options is `TaskSpec`.
5. So we have two options:
a) Have the `_default_container_task_constructor` function receive a `TaskSpec` object.
b) Extend the signature of `_default_container_task_constructor` to pass all `TaskSpec` attributes.
Option b) seems to be more future-proof since we won't need to change the signature again if more options are added to TaskSpec.
If we went with option a) we'd have to extend the signature again to add the caching_strategy. Option b) gives us more stable signature.

What are the main issues you see with this approach?

Reply:
I am opinion-less regarding the implementation of the bespoken component IR here, so I'll not block this PR. Just want to make sure go/component-authoring-tfx-2020  gets refreshed and you got the approval/consensus you need to refactor the corresponding part in TFX, so that this change won't be dangling.

Two things I wish to call out:

1. Things like k8s options/caching/retry etc. sounds like very platform-specific, which is not closely tie to the underlying business logic defined by either the user-specified Python function or ComponentSpec + ExecutorSpec, so I think it's actually a separate work (perhaps something like 'k8s configuration simplication');

2. python func --> ComponentSpec --> factory function --(pass arguments)--> TaskSpec (includes ComponentSpec) + arguments --> BaseComponent object with ExecutorContainerSpec and KubernetesComponentConfig, followed by a step to translate TFX stuff to ContainerOp in KubeflowDagRunner, is a codepath interleaving TFX and KFP stack, which is perhaps not ideal. I think 1) we might need to refactor the code path for a better clarification, and 2) make sure we have a good test coverage. Especially, in the TFX dsl bridge which you're working on, if we indeed depend on the change introduced in this PR, some head-to-head tests might be needed.

WDYT?

Reply:
>Just want to make sure go/component-authoring-tfx-2020 gets refreshed

Sounds good. Although in practice there should not be visible changes there. Maybe just the mention of [BaseComponentConfig](https://github.com/tensorflow/tfx/blob/5ee3de9c77a92e904010c7656d6269d3b28744bf/tfx/orchestration/kubeflow/base_component.py#L69) part of `BaseComponent`.

>Things like k8s options/caching/retry etc. sounds like very platform-specific, which is not closely tie to the underlying business logic defined by either the user-specified Python function or ComponentSpec + ExecutorSpec,

I agree. Especially for the k8s options. These options must not be part of ComponentSpec which should be completely platform-agnostic. But these options need to exist somewhere in the task-like object. In TFX they exist in BaseComponent as [BaseComponentConfig](https://github.com/tensorflow/tfx/blob/5ee3de9c77a92e904010c7656d6269d3b28744bf/tfx/orchestration/kubeflow/base_component.py#L69) part of `BaseComponent`. In `TaskSpec` they are in `task_spec.execution_options`.

> followed by a step to translate TFX stuff to ContainerOp in KubeflowDagRunner, is a codepath interleaving TFX and KFP stack, which is perhaps not ideal.

Note that in the case of the KubeflowDagRunner, the ContainerOp will correspond not to the user component container with possible Kubernetes options, but rather to the KubernetesLauncher. The BaseComponent object with ExecutorContainerSpec and KubernetesComponentConfig will be just packed in JSON and sent to the launcher and won't affect the ContainerOp attributes.

>2) make sure we have a good test coverage. Especially, in the TFX dsl bridge which you're working on,

Sounds good.

>if we indeed depend on the change introduced in this PR, some head-to-head tests might be needed.

It might be easier due to the fact that the bridge CL is not checked in yet, and that KFP SDK can have frequent releases.

I'm a bit torn here.
I do not want to make any more changes to the _default_container_task_constructor signature in the future (since it's a bit like public internal API). But I also understand that this change brings TaskSpec into the picture and this would be another thing that I'll have to explain to the bridge reviewers.

Another alternative is to make the interface have `**kwargs` which sweeps the problem under the carpet...

Reply:
> Another alternative is to make the interface have **kwargs which sweeps the problem under the carpet...

Actually `**kwargs` does not sound very bad to me...

I am wondering if that's acceptable that for short-term, we first enable the users to author python function based component in TFX SDK, with some default options (but still those options can be specified in other ways). Then we solve the configuration simplification part. WDYT?

Reply:
> Actually `**kwargs` does not sound very bad to me...

This looks like a working solution, yes. I guess I'll just use **kwargs in the TFX' bridge so that it does not expose TaskSpec at first. This makes this PR not urgent.

As for KFP we can later have a chat later about configuring per-task Kubernetes and caching options in graph components.
/hold